### DISCUSSION TITLE
Remove ListDirectoryAsync(string path, CancellationToken cancellationToken)

### DISCUSSION THREAD

**Discussion by drieseng** [sshnet/SSH.NET#1113] [General]
@IgorMilavec added a `ListDirectoryAsync(string path, CancellationToken cancellationToken)` to **SftpClient** that returns a **Task<IEnumerable<ISftpFile>**.

I don't want to ship a release (even if it's beta) with this method.
I see two options:
* Remove this method
* Change return type to **IAsyncEnumerable\<ISftpFile>**

Let's discuss these (and perhaps other) options in the next days.

**Comment by WojciechNagorski** [comment]
We have a PR for second opinion but there are problems. Tomorrow I will look at it.
https://github.com/sshnet/SSH.NET/pull/907

**Comment by WojciechNagorski** [comment]
To use **IAsyncEnumerable** on `.NET 4.6.2` and `.NET Standard 2.0` we can add a reference to Microsoft's Nuget [Microsoft.Bcl.AsyncInterfaces](https://www.nuget.org/packages/Microsoft.Bcl.AsyncInterfaces). This package is not required starting with .NET Standard 2.1, .NET Core 3.0 (and higher).

The possible solution:
```csproj
<PropertyGroup>
    <TargetFrameworks>net462;netstandard2.0;net6.0</TargetFrameworks>
    <!-- It is important to use C# 8.0 -->
    <LangVersion>8.0</LangVersion> 
</PropertyGroup>

<ItemGroup Condition="'$(TargetFramework)'=='net462' or '$(TargetFramework)'=='netstandard2.0' ">
    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="7.0.0" />
</ItemGroup>
```

To be honest, I don't like the solution from PR #907. IMO we should choose one of the following options:
1. Expose `IAsyncEnumerable` version of the `ListDirectoryAsync` method using the [Microsoft.Bcl.AsyncInterfaces](https://www.nuget.org/packages/Microsoft.Bcl.AsyncInterfaces) Nuget. There is only one problem. **IAsyncEnumerable** requires `C# 8.0` which is not officially supported in `.NET 4.6.2` and `.NET Standard 2.0`. So users will have to set `<LangVersion>8.0</LangVersion>` if they want to use `ListDirectoryAsync` method. There is more information: https://stackoverflow.com/a/57020770/4809321
2. Add support for `.NET Standard 2.1` and expose `IAsyncEnumerable` version of the `ListDirectoryAsync` method only for .NET 6 and .NET Standard 2.1. For older `.NET 4.6.2` and `.NET Standard 2.0` just remove this method. 

The more I think about it, the more I'm in favor of option 2.

**Reply by drieseng** [reply]
@WojciechNagorski, I also favor the second option.
